README.md
# **_Paratonnerre_**

<p align="center">
<img src="./doc/logo.png" alt= “” width="250" height="250">
</p>

Event-catcher for Hyperledger Fabric.

It connects to a node in the Hyperledger fabric network to listen for events. It has the following features.

- Listening raw block events.
- Listenin filtered block events.
- Listening block and private data events.
- Listening chaincode events.
- Listen to block events on multiple channel.
- Listen to chaincode events on multiple chaincodes.
- Gateway-SDK is used.


## **Installization**

Go to the release page and download docker-compose files.

[Release Page - v0.3.0-alpha](https://github.com/caner-emec/paratonnerre/releases/tag/v0.3.0-alpha)

## **Getting Started**

Before you start, get your Hyperledger Fabric network up and running. Create a channel, upload a chaincode (or multiple chaincodes) and make a few transactions.

### **Create Folder For Testing**

```bash
mkdir paratonnerreTest
cd paratonnerreTest
```

### **Prepare Credentials**

**Copy all three files into "paratonnerreTest" folder.**

In order to run the application, you need to prepare the following 3 files.

- user certificate (**cert.pem**)
- user private key (**\*\_sk**)
- peer tls certificate (**ca.crt**)

</br>

< **For Fabric Samples Test Network** >

You can access the files by following the path below in the test network in the Fabric samples repository.

**For User Certificate:**

- fabric-samples/test-network/organizations/peerOrganizations/org1.example.com/users/User1@org1.example.com/msp/signcerts/**cert.pem**

**For User Private Key:** (After copy-paste, change name to **_priv_sk_**)

- fabric-samples/test-network/organizations/peerOrganizations/org1.example.com/users/User1@org1.example.com/msp/keystore/**< private key file >**

**For Peer Tls Certificate:**

- fabric-samples/test-network/organizations/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/tls/**ca.crt**

### **Get Docker-Compose Files**

Download files and put into paratonnerreTest folder.

[kafka-services compose file](https://github.com/caner-emec/paratonnerre/releases/download/v0.1.0-alpha/kafka-services-docker-compose.yaml)

[paratonnerre compose file](https://github.com/caner-emec/paratonnerre/releases/download/v0.1.0-alpha/paratonnerre-docker-compose.yaml)

### **Configure Environment Variables**

```bash
- PEER_NAME=peer0.org1.example.com
- MSP_ID=Org1MSP
- PEER_ENDPOINT=localhost:7051
- PEER_HOST_ALIAS=peer0.org1.example.com
# Listener settings
- LISTENER_BLOCK_EVENTS_FROM=mychannel # comma seperated channel list
- LISTENER_CHAINCODE_EVENTS_FROM=mychannel:events,mychannel:basic # comma seperated chaincode list, ex. <channel name>:<chaincode name>
```

**Change the above variables according to your own network.**

</br>

### **Start Application**

```bash
docker-compose -f kafka-services-docker-compose.yaml up -d
docker-compose -f paratonnerre-docker-compose.yaml up
```

## **Basic Usage**

**To Listen Events**

Topic names:

- _< KAFKA_TOPIC_HLF_BLOCKS_PREFIX >\_< Channel Name >_
- _< KAFKA_TOPIC_HLF_FILTERED_BLOCK_PREFIX >\_< Channel Name >_
- _< KAFKA_TOPIC_HLF_BLOCK_AND_PRIVDATA_PREFIX >\_< Channel Name >_
- _< KAFKA_TOPIC_HLF_TRANSACTION_PREFIX >\_< Channel Name >\_< Chaincode Name >_

</br>

```bash
docker exec kafka-1 kafka-console-consumer --topic <TOPIC NAME COMES HERE> --from-beginning --bootstrap-server localhost:9092
```

**Examples**

```bash
docker exec kafka-1 kafka-console-consumer --topic hlf_txs_mychannel_events --from-beginning --bootstrap-server localhost:9092
docker exec kafka-1 kafka-console-consumer --topic hlf_txs_mychannel_basic --from-beginning --bootstrap-server localhost:9092
docker exec kafka-1 kafka-console-consumer --topic hlf_blocks_mychannel --from-beginning --bootstrap-server localhost:9092
docker exec kafka-1 kafka-console-consumer --topic hlf_filteredBlock_mychannel --from-beginning --bootstrap-server localhost:9092
docker exec kafka-1 kafka-console-consumer --topic hlf_blockAndPriv_mychannel --from-beginning --bootstrap-server localhost:9092
```

**Block And Private Data Event Example Output**

```json
{
    "hasBlock":true,
    "block":{
        "block":{
            "metadata":{
                "value":"CgIIAhIJCgcKAQEQAhgp",
                "numOfSignatures":1,
                "signatures":[
                    {
                        "signatureHeader":{
                            "creator":{
                                "mspId":"OrdererMSP",
                                "id":"LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUNwakNDQWt5Z0F3SUJBZ0lVZTZHbjhUOVY0ZFRVdzRkSzdZb2oxZ0laN2VRd0NnWUlLb1pJemowRUF3SXcKWWpFTE1Ba0dBMVVFQmhNQ1ZWTXhFVEFQQmdOVkJBZ1RDRTVsZHlCWmIzSnJNUkV3RHdZRFZRUUhFd2hPWlhjZwpXVzl5YXpFVU1CSUdBMVVFQ2hNTFpYaGhiWEJzWlM1amIyMHhGekFWQmdOVkJBTVREbU5oTG1WNFlXMXdiR1V1ClkyOXRNQjRYRFRJek1EVXhPVEV4TlRVd01Gb1hEVEkwTURVeE9ERXlNREF3TUZvd1lERUxNQWtHQTFVRUJoTUMKVlZNeEZ6QVZCZ05WQkFnVERrNXZjblJvSUVOaGNtOXNhVzVoTVJRd0VnWURWUVFLRXd0SWVYQmxjbXhsWkdkbApjakVRTUE0R0ExVUVDeE1IYjNKa1pYSmxjakVRTUE0R0ExVUVBeE1IYjNKa1pYSmxjakJaTUJNR0J5cUdTTTQ5CkFnRUdDQ3FHU000OUF3RUhBMElBQkM5NUU2Nkk4cG9ZblU2QmdnZ2pCcDUvbEx3Y1hkRnFaMCt4TmJWTDJLUk4KT1I4d2VnQWVjUGp6UVRxUHZIREY3T0lqRE5LQTdpVnpqOS9wdTRXTlNmbWpnZUV3Z2Q0d0RnWURWUjBQQVFILwpCQVFEQWdlQU1Bd0dBMVVkRXdFQi93UUNNQUF3SFFZRFZSME9CQllFRkFGOFhtOXdJeHpNYzBpSUdTYVQ3WlQxClJ2WFlNQjhHQTFVZEl3UVlNQmFBRkhySnl6Qm11RURXeWw0R2dWM0JMeUdMa1h5Wk1DRUdBMVVkRVFRYU1CaUMKRm1WNGFXMXBiM1Z6TWkxSVZVMUJMVWcwTFZZMExURXdXd1lJS2dNRUJRWUhDQUVFVDNzaVlYUjBjbk1pT25zaQphR1l1UVdabWFXeHBZWFJwYjI0aU9pSWlMQ0pvWmk1RmJuSnZiR3h0Wlc1MFNVUWlPaUp2Y21SbGNtVnlJaXdpCmFHWXVWSGx3WlNJNkltOXlaR1Z5WlhJaWZYMHdDZ1lJS29aSXpqMEVBd0lEU0FBd1JRSWhBUDc4SlJ2czNFV2EKUFBxTkZTS2laQnhNbzBhRDBRc3dhVFVkZHpkYWdONUpBaUFZeXV2WGdHbXQ2WGxjSm1nbWZwMW95TElHZmV0VwpVYjJ2QUljb0hmcCtmdz09Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K"
                            },
                            "nonce":-9107307609084343000
                        },
                        "signature":"MEQCIHX+jfh6M09x0UOmpu9vMdT6PTEJvUEkoClPs4Gsz97EAiAHkp5dmbhBGfuKrCLqy+nurdK3tfHNG/uZcEYGMqJChw=="
                    }
                ],
                "metadataCommitHash":"0a20383f55343c2cfbdd3a5a2322effea002a4f28af37b1e83f5f43743ab02b9b83a"
            },
            "header":{
                "number":39,
                "previousHash":"8e458f792409ce48fa68cb20208a622ba226747189ff58b44905d62086a33f1c",
                "dataHash":"2e381c6e56a9b82c5da5707a6eea327b9cb942f3d313418d4609d490ef08025f"
            },
            "data":[
                {
                    "txValidationStatus":"VALID",
                    "signature":"30450221008b81d935c066eeaa5d5df715c21575281665c26b93c39833538e08ec16cd5a25022023ec2524c97230c2f4ce12c6ea41bcee5b7cd6c04c12ae94a715f774696cbf68",
                    "payload":{
                        "header":{
                            "header":{
                                "channelHeader":{
                                    "type":"ENDORSER_TRANSACTION",
                                    "version":0,
                                    "timestamp":"2023-06-24T15:37:08.133Z",
                                    "channelId":"mychannel",
                                    "txId":"f2330b37616667a0d9b81221133d090faeff67d9c8a8a81c3c84a17be613569b",
                                    "epoch":0,
                                    "tlsCertHash":""
                                },
                                "signatureHeader":{
                                    "creator":{
                                        "mspId":"Org1MSP",
                                        "id":"LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUNyakNDQWxTZ0F3SUJBZ0lVVmxFZlJVZmhvdk0vUkpMTk9vblBDK2g4Yk93d0NnWUlLb1pJemowRUF3SXcKY0RFTE1Ba0dBMVVFQmhNQ1ZWTXhGekFWQmdOVkJBZ1REazV2Y25Sb0lFTmhjbTlzYVc1aE1ROHdEUVlEVlFRSApFd1pFZFhKb1lXMHhHVEFYQmdOVkJBb1RFRzl5WnpFdVpYaGhiWEJzWlM1amIyMHhIREFhQmdOVkJBTVRFMk5oCkxtOXlaekV1WlhoaGJYQnNaUzVqYjIwd0hoY05Nak13TlRFNU1URTFOVEF3V2hjTk1qUXdOVEU0TVRJd01EQXcKV2pCZE1Rc3dDUVlEVlFRR0V3SlZVekVYTUJVR0ExVUVDQk1PVG05eWRHZ2dRMkZ5YjJ4cGJtRXhGREFTQmdOVgpCQW9UQzBoNWNHVnliR1ZrWjJWeU1ROHdEUVlEVlFRTEV3WmpiR2xsYm5ReERqQU1CZ05WQkFNVEJYVnpaWEl4Ck1Ga3dFd1lIS29aSXpqMENBUVlJS29aSXpqMERBUWNEUWdBRUpNQ1N3RmQ4UlFtOWxBM3JmYlFhM01VVDZPN2IKcU1CU1lzQ09LcEpNczZJUk15T0RCUjRYemhjRkpKeWo0Mkg5U1pVZHllc1ZhZXY2azRLWk80Z0tGcU9CM2pDQgoyekFPQmdOVkhROEJBZjhFQkFNQ0I0QXdEQVlEVlIwVEFRSC9CQUl3QURBZEJnTlZIUTRFRmdRVTVZcm5OaDkvCnA4UWtmSHd4K2xjU2RVbG93SFl3SHdZRFZSMGpCQmd3Rm9BVTdjS1p2S1BqRktSRktkekU2ZUxxNWE4NHcrY3cKSVFZRFZSMFJCQm93R0lJV1pYaHBiV2x2ZFhNeUxVaFZUVUV0U0RRdFZqUXRNVEJZQmdncUF3UUZCZ2NJQVFSTQpleUpoZEhSeWN5STZleUpvWmk1QlptWnBiR2xoZEdsdmJpSTZJaUlzSW1obUxrVnVjbTlzYkcxbGJuUkpSQ0k2CkluVnpaWEl4SWl3aWFHWXVWSGx3WlNJNkltTnNhV1Z1ZENKOWZUQUtCZ2dxaGtqT1BRUURBZ05JQURCRkFpRUEKbnlrQW1VdnhjMWpyRThOdlRVZzErQ0FYM0o5OXY2RGhJYVFkeDdUV2pwY0NJR0xpbS9jeGVjdGlINDBxaXVzbApCa0hqSXRtYzcyU0xQMC84RVd5RHFkWEYKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo="
                                    },
                                    "nonce":-265899332887951070
                                }
                            }
                        },
                        "data":{
                            "transactionActions":[
                                {
                                    "proposalActionHeader":{
                                        "creator":{
                                            "mspId":"Org1MSP",
                                            "id":"LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUNyakNDQWxTZ0F3SUJBZ0lVVmxFZlJVZmhvdk0vUkpMTk9vblBDK2g4Yk93d0NnWUlLb1pJemowRUF3SXcKY0RFTE1Ba0dBMVVFQmhNQ1ZWTXhGekFWQmdOVkJBZ1REazV2Y25Sb0lFTmhjbTlzYVc1aE1ROHdEUVlEVlFRSApFd1pFZFhKb1lXMHhHVEFYQmdOVkJBb1RFRzl5WnpFdVpYaGhiWEJzWlM1amIyMHhIREFhQmdOVkJBTVRFMk5oCkxtOXlaekV1WlhoaGJYQnNaUzVqYjIwd0hoY05Nak13TlRFNU1URTFOVEF3V2hjTk1qUXdOVEU0TVRJd01EQXcKV2pCZE1Rc3dDUVlEVlFRR0V3SlZVekVYTUJVR0ExVUVDQk1PVG05eWRHZ2dRMkZ5YjJ4cGJtRXhGREFTQmdOVgpCQW9UQzBoNWNHVnliR1ZrWjJWeU1ROHdEUVlEVlFRTEV3WmpiR2xsYm5ReERqQU1CZ05WQkFNVEJYVnpaWEl4Ck1Ga3dFd1lIS29aSXpqMENBUVlJS29aSXpqMERBUWNEUWdBRUpNQ1N3RmQ4UlFtOWxBM3JmYlFhM01VVDZPN2IKcU1CU1lzQ09LcEpNczZJUk15T0RCUjRYemhjRkpKeWo0Mkg5U1pVZHllc1ZhZXY2azRLWk80Z0tGcU9CM2pDQgoyekFPQmdOVkhROEJBZjhFQkFNQ0I0QXdEQVlEVlIwVEFRSC9CQUl3QURBZEJnTlZIUTRFRmdRVTVZcm5OaDkvCnA4UWtmSHd4K2xjU2RVbG93SFl3SHdZRFZSMGpCQmd3Rm9BVTdjS1p2S1BqRktSRktkekU2ZUxxNWE4NHcrY3cKSVFZRFZSMFJCQm93R0lJV1pYaHBiV2x2ZFhNeUxVaFZUVUV0U0RRdFZqUXRNVEJZQmdncUF3UUZCZ2NJQVFSTQpleUpoZEhSeWN5STZleUpvWmk1QlptWnBiR2xoZEdsdmJpSTZJaUlzSW1obUxrVnVjbTlzYkcxbGJuUkpSQ0k2CkluVnpaWEl4SWl3aWFHWXVWSGx3WlNJNkltTnNhV1Z1ZENKOWZUQUtCZ2dxaGtqT1BRUURBZ05JQURCRkFpRUEKbnlrQW1VdnhjMWpyRThOdlRVZzErQ0FYM0o5OXY2RGhJYVFkeDdUV2pwY0NJR0xpbS9jeGVjdGlINDBxaXVzbApCa0hqSXRtYzcyU0xQMC84RVd5RHFkWEYKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo="
                                        },
                                        "nonce":-265899332887951070
                                    },
                                    "chaincodeActionPayload":{
                                        "chaincodeEndorsedAction":{
                                            "endorsments":[
                                                {
                                                    "endorser":{
                                                        "mspId":"Org1MSP",
                                                        "id":"LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUNxVENDQWxDZ0F3SUJBZ0lVU2Z4bjRzamYvV2JTUHNOajBpd2VuNmFaSmVVd0NnWUlLb1pJemowRUF3SXcKY0RFTE1Ba0dBMVVFQmhNQ1ZWTXhGekFWQmdOVkJBZ1REazV2Y25Sb0lFTmhjbTlzYVc1aE1ROHdEUVlEVlFRSApFd1pFZFhKb1lXMHhHVEFYQmdOVkJBb1RFRzl5WnpFdVpYaGhiWEJzWlM1amIyMHhIREFhQmdOVkJBTVRFMk5oCkxtOXlaekV1WlhoaGJYQnNaUzVqYjIwd0hoY05Nak13TlRFNU1URTFOVEF3V2hjTk1qUXdOVEU0TVRJd01EQXcKV2pCYk1Rc3dDUVlEVlFRR0V3SlZVekVYTUJVR0ExVUVDQk1PVG05eWRHZ2dRMkZ5YjJ4cGJtRXhGREFTQmdOVgpCQW9UQzBoNWNHVnliR1ZrWjJWeU1RMHdDd1lEVlFRTEV3UndaV1Z5TVE0d0RBWURWUVFERXdWd1pXVnlNREJaCk1CTUdCeXFHU000OUFnRUdDQ3FHU000OUF3RUhBMElBQk9vSk1IeXdBK3Yvcm1jSUNOcE5PZC9wTis5WUQ0WTAKTnl3ZU1VWGxZL3AxdktTZkQrSG0vT3c4QWFFYUVlUzdaZGlrN081SHVmVWtJU2QwY2dlajVBdWpnZHd3Z2RrdwpEZ1lEVlIwUEFRSC9CQVFEQWdlQU1Bd0dBMVVkRXdFQi93UUNNQUF3SFFZRFZSME9CQllFRkJLbWpkU3BVMFk4CkgzY0NWNmszV2laaXlWdy9NQjhHQTFVZEl3UVlNQmFBRk8zQ21ieWo0eFNrUlNuY3hPbmk2dVd2T01Qbk1DRUcKQTFVZEVRUWFNQmlDRm1WNGFXMXBiM1Z6TWkxSVZVMUJMVWcwTFZZMExURXdWZ1lJS2dNRUJRWUhDQUVFU25zaQpZWFIwY25NaU9uc2lhR1l1UVdabWFXeHBZWFJwYjI0aU9pSWlMQ0pvWmk1RmJuSnZiR3h0Wlc1MFNVUWlPaUp3ClpXVnlNQ0lzSW1obUxsUjVjR1VpT2lKd1pXVnlJbjE5TUFvR0NDcUdTTTQ5QkFNQ0EwY0FNRVFDSUFadzZ3S2IKK0dUa1RPYXdaeDBpcE1QWFFzWlk2Q2x6Q05LSkx3QzhWWWtDQWlCc1orTzRqSmhoZFhRZ0t1Z1pIeVMvd2NWUgpVM2Y2bGNORmhNVnRYTFJtenc9PQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg=="
                                                    },
                                                    "signature":"3045022100b2c52a9efcf42d846d4722c27d662f08cca457bab3e31ae115b2deccfac49bfa0220178f9997609f70de54ab9cc52e15447b7f9fae930aa903d06f41257269ecb7ea"
                                                }
                                            ],
                                            "proposalResponsePayload":{
                                                "proposalHash":"f435152bc56d8ab7296495a1cee15dd556d75c46e520c2a37ef74adf06ea00a5",
                                                "extensionChaincodeAction":{
                                                    "chaincodeId":{
                                                        "name":"private",
                                                        "path":"",
                                                        "version":"1.0"
                                                    },
                                                    "chaincodeEvent":{
                                                        "chaincodeId":"",
                                                        "txId":"",
                                                        "eventName":"",
                                                        "payload":""
                                                    },
                                                    "chaincodeResponse":{
                                                        "status":200,
                                                        "message":""
                                                    },
                                                    "chaincodeResults":{
                                                        "txRWSet":{
                                                            "datamodel":0,
                                                            "nsRWSet":[
                                                                {
                                                                    "namespace":"_lifecycle",
                                                                    "rwset":{
                                                                        "kvReadList":[
                                                                            {
                                                                                "key":"namespaces/fields/private/Collections",
                                                                                "version":{
                                                                                    "blockNum":34,
                                                                                    "txNum":0
                                                                                }
                                                                            },
                                                                            {
                                                                                "key":"namespaces/fields/private/EndorsementInfo",
                                                                                "version":{
                                                                                    "blockNum":34,
                                                                                    "txNum":0
                                                                                }
                                                                            },
                                                                            {
                                                                                "key":"namespaces/fields/private/Sequence",
                                                                                "version":{
                                                                                    "blockNum":34,
                                                                                    "txNum":0
                                                                                }
                                                                            },
                                                                            {
                                                                                "key":"namespaces/fields/private/ValidationInfo",
                                                                                "version":{
                                                                                    "blockNum":34,
                                                                                    "txNum":0
                                                                                }
                                                                            },
                                                                            {
                                                                                "key":"namespaces/metadata/private",
                                                                                "version":{
                                                                                    "blockNum":34,
                                                                                    "txNum":0
                                                                                }
                                                                            }
                                                                        ],
                                                                        "kvWriteList":[
                                                                            
                                                                        ],
                                                                        "rangeQueryInfoList":[
                                                                            
                                                                        ],
                                                                        "kvMetadataWriteList":[
                                                                            
                                                                        ]
                                                                    },
                                                                    "collectionHashedRWSet":[
                                                                        
                                                                    ]
                                                                },
                                                                {
                                                                    "namespace":"private",
                                                                    "rwset":{
                                                                        "kvReadList":[
                                                                            
                                                                        ],
                                                                        "kvWriteList":[
                                                                            
                                                                        ],
                                                                        "rangeQueryInfoList":[
                                                                            
                                                                        ],
                                                                        "kvMetadataWriteList":[
                                                                            
                                                                        ]
                                                                    },
                                                                    "collectionHashedRWSet":[
                                                                        {
                                                                            "collectionName":"Org1MSPPrivateCollection",
                                                                            "hashedRWSet":{
                                                                                "hashedReadList":[
                                                                                    
                                                                                ],
                                                                                "hashedWriteList":[
                                                                                    {
                                                                                        "keyHAsh":"d53762bfc6f701bfb680d6f8b334e416b5f77f0ee1eae9d7a2bf856ac2da46a4",
                                                                                        "valueHash":"44e1b27f5ed029718dd4a7668b493cc714515c275d59034d273c1d4ad03bfca7",
                                                                                        "isDelete":false,
                                                                                        "isPurge":false
                                                                                    }
                                                                                ],
                                                                                "metadataWriteList":[
                                                                                    
                                                                                ]
                                                                            },
                                                                            "privateRWSetHash":"12440a20d53762bfc6f701bfb680d6f8b334e416b5f77f0ee1eae9d7a2bf856ac2da46a41a2044e1b27f5ed029718dd4a7668b493cc714515c275d59034d273c1d4ad03bfca7"
                                                                        },
                                                                        {
                                                                            "collectionName":"assetCollection",
                                                                            "hashedRWSet":{
                                                                                "hashedReadList":[
                                                                                    {
                                                                                        "keyHash":"d53762bfc6f701bfb680d6f8b334e416b5f77f0ee1eae9d7a2bf856ac2da46a4",
                                                                                        "version":{
                                                                                            
                                                                                        }
                                                                                    }
                                                                                ],
                                                                                "hashedWriteList":[
                                                                                    {
                                                                                        "keyHAsh":"d53762bfc6f701bfb680d6f8b334e416b5f77f0ee1eae9d7a2bf856ac2da46a4",
                                                                                        "valueHash":"681a7da202d907bb6fa3cffaeac0424ec7819b696a71cbe997d0bd0ee25a35e6",
                                                                                        "isDelete":false,
                                                                                        "isPurge":false
                                                                                    }
                                                                                ],
                                                                                "metadataWriteList":[
                                                                                    
                                                                                ]
                                                                            },
                                                                            "privateRWSetHash":"0a220a20d53762bfc6f701bfb680d6f8b334e416b5f77f0ee1eae9d7a2bf856ac2da46a412440a20d53762bfc6f701bfb680d6f8b334e416b5f77f0ee1eae9d7a2bf856ac2da46a41a20681a7da202d907bb6fa3cffaeac0424ec7819b696a71cbe997d0bd0ee25a35e6"
                                                                        }
                                                                    ]
                                                                }
                                                            ]
                                                        },
                                                        "txPvtRWSet":{
                                                            "datamodel":0,
                                                            "nsPvtRWSet":[
                                                                {
                                                                    "namespace":"_lifecycle",
                                                                    "collectionPvtRWSet":[
                                                                        {
                                                                            "collectionName":"\n\u001bnamespaces/metadata/private\u0012\u0002\b\"",
                                                                            "rwset":{
                                                                                "kvReadList":[
                                                                                    
                                                                                ],
                                                                                "kvWriteList":[
                                                                                    
                                                                                ],
                                                                                "rangeQueryInfoList":[
                                                                                    
                                                                                ],
                                                                                "kvMetadataWriteList":[
                                                                                    
                                                                                ]
                                                                            }
                                                                        }
                                                                    ]
                                                                },
                                                                {
                                                                    "namespace":"private",
                                                                    "collectionPvtRWSet":[
                                                                        
                                                                    ]
                                                                }
                                                            ]
                                                        }
                                                    }
                                                }
                                            }
                                        },
                                        "chaincodeProposalPayload":{
                                            "chaincodeInvocationSpecInput":{
                                                "ChaincodeSpec":{
                                                    "type":"NODE",
                                                    "chaincodeId":{
                                                        "name":"private",
                                                        "path":"",
                                                        "version":""
                                                    },
                                                    "input":{
                                                        "args":[
                                                            "CreateAsset"
                                                        ],
                                                        "isInit":false
                                                    },
                                                    "timeout":0
                                                }
                                            }
                                        }
                                    }
                                }
                            ]
                        }
                    }
                }
            ]
        }
    },
    "privateDataMap":[
        [
            {
                "namespace":"private",
                "collectionPvtRWSet":[
                    {
                        "collectionName":"Org1MSPPrivateCollection",
                        "rwset":{
                            "kvReadList":[
                                
                            ],
                            "kvWriteList":[
                                {
                                    "key":"asset1687621028121",
                                    "value":{
                                        "asByte":{
                                            "0":123,
                                            "1":34,
                                            "2":65,
                                            "3":112,
                                            "4":112,
                                            "5":114,
                                            "6":97,
                                            "7":105,
                                            "8":115,
                                            "9":101,
                                            "10":100,
                                            "11":86,
                                            "12":97,
                                            "13":108,
                                            "14":117,
                                            "15":101,
                                            "16":34,
                                            "17":58,
                                            "18":49,
                                            "19":48,
                                            "20":48,
                                            "21":44,
                                            "22":34,

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.3.0-alpha](https://github.com/caner-emec/paratonnerre/compare/v0.1.0-alpha...v0.3.0-alpha) (2023-06-24)

### Features

- feature: Block and private data listener added! ([a10cd7b](https://github.com/caner-emec/paratonnerre/commit/a10cd7bf1de58a0a5810d40e8dea5e54bc0f977e))
- filtered block event support added! ([77c382d](https://github.com/caner-emec/paratonnerre/commit/77c382d1b90dc5d5a633b0fba8e33797549a8b32))

## 0.1.0-alpha (2023-06-11)

### Features

- added block parser and types etc ([efb395a](https://github.com/caner-emec/paratonnerre/commit/efb395ab011ebf474e6df9d9efb8782e460570dc))
- added kafka brokers and topics ([6783ad1](https://github.com/caner-emec/paratonnerre/commit/6783ad170faa5ee9ce6ff8a6400bd166bb7ded0a))
- added sample data files ([cdfd181](https://github.com/caner-emec/paratonnerre/commit/cdfd1818db3f92af59084930f07b5c487ca1e7e2))
- block parser added! ([f41d20c](https://github.com/caner-emec/paratonnerre/commit/f41d20c027d87ac851bfe29e72cf742eed78b874))
- chaincode event listener added! ([fd6ad70](https://github.com/caner-emec/paratonnerre/commit/fd6ad7008cbb4be04cfe51d398c90c3ce81d11fe))
- checkpointer added! ([96fa6a3](https://github.com/caner-emec/paratonnerre/commit/96fa6a3585a0b53d91435df53ff62aa09fdc33f8))
- kafka dynamic topic creation added! ([57a8caa](https://github.com/caner-emec/paratonnerre/commit/57a8caac04eef1e60120378d215f2410a9bbebf1))
- logging added ([481bae0](https://github.com/caner-emec/paratonnerre/commit/481bae0b3dfabffcb237cde74d9530695a12f773))
- multiple listeners utility added! ([3582c6c](https://github.com/caner-emec/paratonnerre/commit/3582c6c3d53fb382bd7196a71d73efb7ea6796fa))


src/types/block.types.ts
import {common} from '@hyperledger/fabric-protos';
import {
  ProcessedBlockHeader,
  ProcessedBlockMetadataSignature,
  ProcessedHeaderTypeEnum,
  ProcessedId,
  ProcessedSignatureHeader,
} from './default.types';

type ProcessedEnvelope = {
  txValidationStatus: string | TxValidationCode;
  signature: string;
  payload: ProcessedPayload;
};

// TODO!! header and data type
type ProcessedPayload = {
  header: object;
  data: object;
};

type ProcessedPayloadHeader = {
  header: {
    channelHeader: {
      type: ProcessedHeaderTypeEnum | string;
      version: number;
      timestamp: Date | undefined;
      channelId: string;
      txId: string;
      epoch: number;
      tlsCertHash: string;
    };
    signatureHeader: ProcessedSignatureHeader;
  };
};

type ProcessedEndorsmentEntry = {
  endorser: ProcessedId;
  signature: string;
};

type ProcessedBlockMetadata = {
  value: string | Uint8Array;
  numOfSignatures: number;
  signatures: ProcessedBlockMetadataSignature[];
  metadataCommitHash: string;
};

type ProcessedBlockConstruction = {
  block: {
    metadata: ProcessedBlockMetadata;
    header: ProcessedBlockHeader;
    data: ProcessedEnvelope[];
  };
};

enum TxValidationCode {
  VALID = 0,
  NIL_ENVELOPE = 1,
  BAD_PAYLOAD = 2,
  BAD_COMMON_HEADER = 3,
  BAD_CREATOR_SIGNATURE = 4,
  INVALID_ENDORSER_TRANSACTION = 5,
  INVALID_CONFIG_TRANSACTION = 6,
  UNSUPPORTED_TX_PAYLOAD = 7,
  BAD_PROPOSAL_TXID = 8,
  DUPLICATE_TXID = 9,
  ENDORSEMENT_POLICY_FAILURE = 10,
  MVCC_READ_CONFLICT = 11,
  PHANTOM_READ_CONFLICT = 12,
  UNKNOWN_TX_TYPE = 13,
  TARGET_CHAIN_NOT_FOUND = 14,
  MARSHAL_TX_ERROR = 15,
  NIL_TXACTION = 16,
  EXPIRED_CHAINCODE = 17,
  CHAINCODE_VERSION_CONFLICT = 18,
  BAD_HEADER_EXTENSION = 19,
  BAD_CHANNEL_HEADER = 20,
  BAD_RESPONSE_PAYLOAD = 21,
  BAD_RWSET = 22,
  ILLEGAL_WRITESET = 23,
  INVALID_WRITESET = 24,
  INVALID_CHAINCODE = 25,
  NOT_VALIDATED = 254,
  INVALID_OTHER_REASON = 255,
}

export {
  ProcessedEnvelope,
  ProcessedPayload,
  ProcessedPayloadHeader,
  ProcessedEndorsmentEntry,
  TxValidationCode,
  ProcessedBlockMetadata,
  ProcessedBlockConstruction,
};
